src/assignments/ray/lesson_3_answers.clj

```diff
@@ -48,7 +48,156 @@
+        count-val-count-ranks (count (vals (frequencies rank)))]
+    ;; printing out just for testing
+    (println "")
+    (println "Hello this is a new hand")
+    (println suit)
+    (println rank)
+    (println count-suits)
+    (println val-count-suits)
+    (println count-suits-count)
+    (println count-ranks)
+    (println count-ranks-count)
+    (println "val-count-ranks" val-count-ranks)
+    (println "count-val-count-ranks" count-val-count-ranks)
+    (cond
+      ;; conditions explained, further differentiate will be done in the next iteration
+      (= count-ranks-count 5) :flush-or-straight-flush
+      (= count-ranks-count 4) :pair
+      (and (= count-ranks-count 3) (and (= count-val-count-ranks 3) (= (first val-count-ranks) 2))) :two-pair
+      (and (= count-ranks-count 3) (and (= count-val-count-ranks 3) (= (first val-count-ranks) 3))) :three-of-a-kind
+      (and (= count-ranks-count 2) (and (= count-val-count-ranks 2) (= (first val-count-ranks) 4))) :four-of-a-kind
+      (and (= count-ranks-count 2) (and (= count-val-count-ranks 2) (= (first val-count-ranks) 3))) :full-house
+      (= count-suits-count 1) :flush-or-straight-flush)))
+
+;; test for sanity
+(= (best-hand-3 ["DQ" "HK" "HA" "C9" "DJ"]) :flush-or-straight-flush)
+(= (best-hand-3 ["DQ" "HQ" "HA" "C9" "DJ"]) :pair)
+(= (best-hand-3 ["DQ" "HQ" "CA" "HA" "DJ"]) :two-pair)
+(= (best-hand-3 ["SQ" "DQ" "HQ" "C9" "DJ"]) :three-of-a-kind)
+(= (best-hand-3 ["SK" "DK" "HK" "CK" "DJ"]) :four-of-a-kind)
+(= (best-hand-3 ["SJ" "DJ" "HJ" "D9" "H9"]) :full-house)
+(= (best-hand-3 ["DQ" "DK" "DA" "D9" "DJ"]) :flush-or-straight-flush)
+
+;; Fourth iteration: All conditions mapped
+
+(defn subseq? [sub] 
+  (some #{sub} (partition 5 1 (map vector "A23456789TJQKA"))))
+
+(subseq? ["2" "3" "4" "5" "6"])
+
+(defn subseqq? [a b]
+  (some #{a} (partition (count a) 1 b)))
+
+(subseqq? [1 2 3] [1 2 3 4 5 6])
+
+(map vector "A1234")
+
+(defn best-hand-4 [hand]
+  ;; getting suits and ranks counts for initial hand determination
+  (let [[suit rank] (apply map vector hand)
+        count-suits (frequencies suit)
+        val-count-suits (vals (frequencies suit))
+        count-suits-count (count count-suits)
+        count-ranks (frequencies rank)
+        count-ranks-count (count count-ranks)
+        val-count-ranks (vals (frequencies rank))
+        count-val-count-ranks (count (vals (frequencies rank)))]
+    ;; printing out just for testing
+    (println "")
+    (println "Hello this is a new hand-4")
+    (println suit)
+    (println rank)
+    (println count-suits)
+    (println "val-count-suits" val-count-suits)
+    (println count-suits-count)
+    (println count-ranks)
+    (println count-ranks-count)
+    (println "val-count-ranks" val-count-ranks)
+    (println "count-val-count-ranks" count-val-count-ranks)
+    (cond
+      ;; conditions explained, further differentiate will be done in the next iteration
+      (= count-ranks-count 5) :straight-flush
```

Comment:
I'm hopelessly lost in the counts and the ranks and the vals and the count-ranks and the rank-counts.

I'm afraid this code is pretty unreadable... :-(

Reply:
Yeah, sorry Alain, the prints were not meant to be readable, it was just used by me printing the result out for testing purposes. I realised this wasn't a great idea to be included in commits so I never include them in any of my subsequent commits.